docs/src/sections/MenuItemSection.js

```diff
@@ -22,7 +22,7 @@ export default function MenuItemSection() {
         <li><code>divider</code>: Adds an horizontal divider between sections</li>
         <li><code>disabled</code>: shows the item as disabled, and prevents the onclick</li>
         <li><code>eventKey</code>: passed to the callback</li>
-        <li><code>onSelect</code>: a callback that is called when the user clicks the item.</li>
+        <li><code>onSelect</code>: a callback that is called when the user clicks the item. <b>NOTE:</b> the containing DropdownMenu actually re-writes this property on its direct child elements. For simple (direct) uses of MenuItem, this will not affect you. However, if you wrap a MenuItem in your own component, be careful! You <em>must</em> handle the onSelect property accordingly, or the <a href="https://github.com/react-bootstrap/react-bootstrap/issues/2429">dropdown will not close</a> after a user selects a menu item!</li>
```

Comment:
You're looking at the wrong API here, what the dropdown does isn't really relevant to your custom MenuItem, only the API of `MenuItem` is important. Phrased another way, the DropdownMenu expects a component that implements the same API as a MenuItem. If you break that expectation then yes, it doesn't work as expected. So no you don't have have to plan on any prop, being provided, just the ones defined in the MenuItem propTypes (which is sort of like an interface). , 

 If you wrap MenuItem then you need to at a _minimum_ provide the same API to expect it to act the same. In this case, MenuItem requires an onSelect prop in order to communicate to it's parent when it's been selected. You are free to ignore or change that expectation by wrapping and not calling the onSelect handler, but then no parent component knows the MenuItem is selected as such doesn't do anything. That's all not an implementation detail is the explicitly defined API contract between the two components. You have the power to change that by implementing a MenuItem that ignores onSelect. Like i said this is a feature, it meanas users can really control how these component behave, whereas if we did it differently you'd have no ability to prevent selection of a menu item.